include/deal.II/trilinos/nox.templates.h

```diff
@@ -0,0 +1,1027 @@
+    , parameters(parameters)
+    , n_residual_evaluations(0)
+    , n_jacobian_applications(0)
+    , n_nonlinear_iterations(0)
+    , n_last_linear_iterations(0)
+  {}
+
+
+
+  template <typename VectorType>
+  void
+  NOXSolver<VectorType>::clear()
+  {
+    // clear interal counters
+    n_residual_evaluations   = 0;
+    n_jacobian_applications  = 0;
+    n_nonlinear_iterations   = 0;
+    n_last_linear_iterations = 0;
+  }
+
+
+
+  template <typename VectorType>
+  unsigned int
+  NOXSolver<VectorType>::solve(VectorType &solution)
+  {
+    if (additional_data.reuse_solver == false)
+      clear(); // clear state
+
+    // create group
+    const auto group = Teuchos::rcp(new internal::NOXWrappers::Group<
+                                    VectorType>(
+      solution,
+      [&](const VectorType &x, VectorType &f) -> int {
+        Assert(
+          residual,
+          ExcMessage(
+            "No residual function has been attached to the NOXSolver object."));
+
+        n_residual_evaluations++;
+
+        // evalute residual
+        return residual(x, f);
+      },
+      [&](const VectorType &x) -> int {
+        Assert(
+          setup_jacobian,
+          ExcMessage(
+            "No setup_jacobian function has been attached to the NOXSolver object."));
+
+        // setup Jacobian
+        int flag = setup_jacobian(x);
+
+        if (flag != 0)
+          return flag;
+
+        if (setup_preconditioner)
+          {
+            // check if preconditioner needs to be updated
+            bool update_preconditioner =
+              ((additional_data.threshold_nonlinear_iterations > 0) &&
+               ((n_nonlinear_iterations %
+                 additional_data.threshold_nonlinear_iterations) == 0)) ||
+              (solve_with_jacobian_and_track_n_linear_iterations &&
+               (n_last_linear_iterations >
+                additional_data.threshold_n_linear_iterations));
+
+            if ((update_preconditioner == false) &&
+                (update_preconditioner_predicate != nullptr))
+              update_preconditioner = update_preconditioner_predicate();
+
+            if (update_preconditioner) // update preconditioner
+              flag = setup_preconditioner(x);
+          }
+
+        return flag;
+      },
+      [&](const VectorType &x, VectorType &v) -> int {
+        Assert(
+          apply_jacobian,
+          ExcMessage(
+            "No apply_jacobian function has been attached to the NOXSolver object."));
+
+        n_jacobian_applications++;
+
+        // apply Jacobian
+        return apply_jacobian(x, v);
+      },
+      [&](const VectorType &f, VectorType &x, const double tolerance) -> int {
+        n_nonlinear_iterations++;
+
+        // invert Jacobian
+        if (solve_with_jacobian)
+          {
+            // without tracking of linear iterations
+            return solve_with_jacobian(f, x, tolerance);
+          }
+        else if (solve_with_jacobian_and_track_n_linear_iterations)
```

Comment:
If I understand correctly, if I set both `solve_with_jacobian` and `solve_with_jacobian_and_track_n_linear_iterations` then only the former would be executed. This is fine, but I would think that we'd want to `Assert(!solve_with_jacobian_and_track_n_linear_iterations,...)` in the first condition. Not being able to get out the `n_linear_iterations` is probably not what the user would expect in such a case. Alternatively, one could permute the two conditions so that the one that tracks n_linear_iterations takes precedence.